runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -722,9 +732,62 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
             TR::Node *replacement = acmpHandler(this, node);
             TR_ASSERT_FATAL_WITH_NODE(node, replacement == node, "can't replace n%un here",
                   node->getGlobalIndex());
+
+            transformObjectCmp = true;
             }
          }
-      else
+      else if ((isLhsValue == TR_yes) && (isRhsValue == TR_yes) && (lhsClass == rhsClass) && lhsClass)
+         {
+         const TR::TypeLayout *fieldTypeLayout = comp()->typeLayout(lhsClass);
+         size_t fieldCount = fieldTypeLayout->count();
+
+         if (fieldCount == 0)
+            {
+            transformObjectCmp = true;
+            }
+         else if (fieldCount == 1)
+            {
+            const TR::TypeLayoutEntry &fieldEntry = fieldTypeLayout->entry(0);
+            TR::DataType dataType = fieldEntry._datatype;
+
+            if (dataType.isIntegral())
+               {
+               transformObjectCmp = true;
+               }
+            else if (dataType.isAddress())
+               {
+               TR_OpaqueClassBlock *fieldClazz = fe()->getClassFromSignature(fieldEntry._typeSignature,
+                                                                             (int32_t)strlen(fieldEntry._typeSignature),
+                                                                             comp()->getCurrentMethod());
+               // If the value type class has one field which is a type of identity class,
+               // the helper call can be replaced with acmp
+               if (fieldClazz && (TR::Compiler->cls.isValueTypeClass(fieldClazz) == TR_no))
+                  {
+                  transformObjectCmp = true;
+                  }
+               }
+            }
+         else if (comp()->cg()->getSupportsArrayCmp() &&
+                  TR::Compiler->cls.classSupportsDirectMemoryComparison(lhsClass))
```

Comment:
Yes, concurrent scavenge. The condition you've posted is more or less the one used in codegen to decide whether to replace an indirect load with `ardbari`:
https://github.com/eclipse-openj9/openj9/blob/8c1840d808785e26daf7eac6096d5c6be17f1c3e/runtime/compiler/codegen/J9CodeGenerator.cpp#L360-L367

I'm not sure why codegen tests `isZ()` as well. I don't think we have another GC mode that requires read barriers. I would expect that since concurrent scavenge is only enabled on z, the other platforms would just always have read barrier type `gc_modron_readbar_none`

I suppose though that `classSupportsDirectMemoryComparison()` is the one from #14885, i.e. that it just checks `J9ClassCanSupportFastSubstitutability`. It looks like any reference-typed field will disqualify a value from `memcmp()`, which makes the question moot at the moment:
https://github.com/eclipse-openj9/openj9/blob/8c1840d808785e26daf7eac6096d5c6be17f1c3e/runtime/vm/createramclass.cpp#L1989-L1992

I was under the impression that reference-typed fields would be allowed, since (outside of concurrent scavenge, and I think even under concurrent scavenge when no scavenge is ongoing) `memcmp()` will work on them. And it looks like the VM calls [a method that handles concurrent scavenge anyway](https://github.com/eclipse-openj9/openj9/blob/8c1840d808785e26daf7eac6096d5c6be17f1c3e/runtime/gc_include/ObjectAccessBarrierAPI.hpp#L397-L413), so I don't know why `L`-types are disqualifying

Reply:
Thanks @jdmpapin! 
@hzongaro Since `J9ClassCanSupportFastSubstitutability` is not set if there is any reference-typed field, should we still need add the check `TR::Compiler->om.readBarrierType() != gc_modron_readbar_none` here?

Reply:
Just a few lines down from the code that Devin quoted, there is code that allows fields of array types to participate in that "fast substitutability" optimization:

https://github.com/eclipse-openj9/openj9/blob/8c1840d808785e26daf7eac6096d5c6be17f1c3e/runtime/vm/createramclass.cpp#L1996-L1998

Assuming that the references that those sorts of fields contain could run into the same issue with pause-not-so-much garbage collection, I think that would mean that the check of `TR::Compiler->om.readBarrierType() != gc_modron_readbar_none` would still be needed.

I think we need someone who is familiar with the rules of that GC policy to answer the question for certain.

Reply:
@tajila I'm wondering if you could shed some light on if checking `J9ClassCanSupportFastSubstitutability` is good enough for memcmp of two VTs. Should we also add checking on if it is concurrent scavenge (`TR::Compiler->om.readBarrierType() != gc_modron_readbar_none`) and if `J9ClassHasReferences` is not set?

Reply:
`J9ClassCanSupportFastSubstitutability` is sufficient, currently it will not be set if there is a reference (any reference) and if there is a floating point member.

Reply:
> currently it will not be set if there is a reference (any reference)

@tajila Except it looks like `J9ClassCanSupportFastSubstitutability` will be set if one of the fields is a reference to an array.  Is that a problem?

Reply:
Ah yes, arrays are fine because they are reference types we can prove are final (cant be sublcasses by a VT) just by looking at the signature. In all other cases more analysis is required.

Reply:
Addressed in 06a2cc2 by updating `classSupportsDirectMemoryComparison` to check if the class has reference field and the GC policy is concurrent scavenger, similar to [the VM check](https://github.com/eclipse-openj9/openj9/blob/8c1840d808785e26daf7eac6096d5c6be17f1c3e/runtime/gc_include/ObjectAccessBarrierAPI.hpp#L397-L413)